gapis/api/templates/api.go.tmpl

```diff
@@ -396,21 +589,55 @@ import (
   func (c {{$name}}) IsNil() bool { return c.data == nil }
 
   // SetNil nullifies this reference.
-  func (c {{$name}}) SetNil() { c.data = nil }
+  func (c {{$name}}) SetNil() {
+    c.data = nil
+    c.ϟrefID = ϟapi.NilRefID
+  }
 
   // Clone makes a deep copy of this reference.
   func (c {{$name}}) Clone(ϟa arena.Arena, ϟseen ϟapi.CloneContext) {{$name}} {
     if c.IsNil() { return Nil{{$name}} }
-    if existing, ok := ϟseen[c.data]; ok {
-      return {{$name}}{data: existing.(*{{$data}})}
+    if existing, ok := ϟseen[c.ϟrefID]; ok {
+      return existing.({{$name}})
     }
-    d := &{{$data}}{}
-    ϟseen[c.data] = d
```

Comment:
This was here on purpose - before the recursion - to handle cycles.

Reply:
Understood. I have changed the `Clone` templates so that `ϟseen` is set before the recursive calls.